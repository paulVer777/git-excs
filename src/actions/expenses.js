import uuidv4 from 'uuid/v4'

const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0

    } = {}) => (
    {
     type:'ADD_EXPENSE',
     expense:{
       id:uuidv4(),
       description,
       note,
       amount,
       createdAt
     }
    }
)

const removeExpense = ({id} = {}) => {

    return {
        type:'REMOVE_EXPENSE',
        id
    }

}
    


const editExpense = (id, update) => (

    {
        type:'EDIT_EXPENSE',
        id,
        update
    } 
)

  


    
export {addExpense,removeExpense,editExpense}