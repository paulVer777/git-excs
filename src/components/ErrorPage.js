import React from 'react'
import {Link} from 'react-router-dom'

const ErrorPage = () => (
    <div>Error Page <Link to='/'>Home</Link></div>
)

export default ErrorPage