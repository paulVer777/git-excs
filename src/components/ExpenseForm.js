import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


export default class ExpenseForm extends React.Component {

constructor(props){
    super(props)
    this.state = {
     description: props.expense ? props.expense.description : '',
     note: props.expense ? props.expense.note : '',
     amount: props.expense ? (props.expense.amount / 100).toString() : '',
     createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
     calendarFocused:false,
     error:undefined
    }
}

setDescription = (event) => { // this = parent scope

    const description = event.target.value
    this.setState((prevState) => ({description}))

}
setNote = (event) => {
    const note = event.target.value
    this.setState((prevState) => ({note}))
}
setAmount = (event) => {
    const amount = event.target.value

    // amount.match(/^\d{1,}(\.\d{0,2})?$/) && this.setState((prevState) => ({amount}))
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState((prevState) => ({amount}))
    }
}
onDateChange = (createdAt) => { // that function takes as a argument moment object
    this.setState((prevState) => ({createdAt}))
}
onFocusChange = ({focused}) =>{
 this.setState((prevState) => ({ calendarFocused:focused }))
}
onSubmitHandler = (event) => {
    event.preventDefault()
    if(!this.state.description || !this.state.amount ){
    this.setState((prevState) => ({error:'Please provide description and amount'}))
    }
    else{
    this.setState((prevState) => ({error:undefined}))
    
    this.props.onSubmit({
        description:this.state.description,
        note:this.state.note,
        amount:parseFloat(this.state.amount),
        createdAt:this.state.createdAt.valueOf()
    })
    }
}

render(){
    return (
        <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit = {this.onSubmitHandler}>
        <input type ='text' placeholder ='Description' autoFocus value = {this.state.description}
        onChange ={this.setDescription}
        />
        <input type ='text' placeholder ='amount'
        value = {this.state.amount}
        onChange ={this.setAmount}
        />
        <SingleDatePicker 
        date = {this.state.createdAt}
        onDateChange = {this.onDateChange}
        focused = {this.state.calendarFocused}
        onFocusChange = {this.onFocusChange}
        numberOfMonths = {1}
        isOutsideRange = {() => false} // all days will be avaivable to choose, including past
        />
        <textarea placeholder ='Add a note to your expense' value ={this.state.note}
        onChange ={this.setNote}
        />
        <button>Add expense</button>
        </form>
        </div>
    )}
}

