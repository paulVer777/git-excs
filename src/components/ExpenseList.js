import React from 'react'
import {connect} from 'react-redux' //connects component to the redux store
import ListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = (props) => ( // {...value } it destructures the props object so we pass variables instead of whole object , it looks like - description:'asdad' amount:'fdsff'
// dispatch    
<div> 
     
    {props.expenses.length === 0 ? (

        <p>You dont have any expenses yet</p>
    )
    :
    (
        props.expenses.map((value,index) => <ListItem key={value.id} {...value} />
    )
    )
    }
    </div>
)

// connect returns a function , and we need to call this (thats why we have "() after () " ) function and pass our component in as a argument
// In connect function we need to pass information we want to connect, we dont need the whole store, we just need a subset. 
// so we pass function as a argument and this function lets us determine what information from the store we want to grab 
// in that function we pass state as a argument and return an object on this object we can put any key value pairs we like 
// usually they are going to be things from the state (everything we put on this object will be accesible to a component we want to connect with 
//( through props))



const mapStateToProps = (state) => ( //when the store changes this callback function automatically rerun providing fresh values to us
    {                                // when you connect a component to the redux store its reactive which means as the store changes
        expenses:getVisibleExpenses(state.expenses,state.filter)     // our component is going to re render new values to the screen
    }
)

export default connect(mapStateToProps)(ExpenseList)


