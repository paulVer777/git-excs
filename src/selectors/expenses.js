import moment from 'moment'

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
 
    return expenses.filter((value,index) => {
   
      const createdAtMoment = moment(value.createdAt)
      
      const startDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true
      const endDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day') : true
      // const startDateMatch = typeof startDate !== 'number' || value.createdAt >= startDate
      // const endDateMatch = typeof endDate !== 'number' || value.createdAt <= endDate
      const textMatch = value.description.toLowerCase().includes(text.toLowerCase())
   
      return startDateMatch && endDateMatch && textMatch // checks if item should pass criteria and be returned
   
    }).sort((a,b) => {

      if(sortBy === 'date'){
        return a.createdAt > b.createdAt ? -1 : 1
      }
      else if(sortBy === 'amount'){
        return a.amount > b.amount ? -1 : 1
      }
    })
   }
   
   export default getVisibleExpenses


   //this function gets actual state and returns new array base of filter object values.