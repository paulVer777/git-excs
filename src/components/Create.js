import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpense} from '../actions/expenses'

export class Create extends React.Component {

    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push(`/`)
    }

    render(){
        return (
    <div>
        <h1>Create page</h1>
        <ExpenseForm
            onSubmit = {this.onSubmit}
        />
    </div>
        )
    }
}
    
const mapDispatchToProps = (dispatch) => ({ //we pulled out the functions from components and passed it here in order to avoid redefining these functions every time component renders

    addExpense:(expense) => dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(Create)