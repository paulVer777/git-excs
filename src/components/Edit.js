import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {editExpense} from '../actions/expenses'
import {removeExpense} from '../actions/expenses'

export class Edit extends React.Component { 

    onSubmit = (expenseObj) => {

        this.props.editExpense(expenseObj,this.props.expense.id)
        this.props.history.push('/')
    }
    
    onClickHandlerRemove = () => {
        this.props.removeExpense(this.props.match.params.id)
        this.props.history.push('/')
    }

    render(){

        return(
            <div>
                <ExpenseForm
    
            expense = {this.props.expense}
            onSubmit = {this.onSubmit}
    
                />
                <button onClick = {this.onClickHandlerRemove}> Remove </button>     
        </div>
            )
    }
}
   
const mapDispatchToProps = (dispatch) => ({
    editExpense:( expenseObj,id ) => dispatch(editExpense(id,expenseObj)),
    removeExpense: (id) => dispatch(removeExpense({id}))

})

const mapStateToProps = (state,props) => ({
    expense: state.expenses.find((value,index) => value.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(Edit)