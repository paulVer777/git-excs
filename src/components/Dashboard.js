import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const Dashboard = (props) => (
    <div>
    <ExpenseListFilters/>
    <ExpenseList/>
    
    </div>
)

export default Dashboard