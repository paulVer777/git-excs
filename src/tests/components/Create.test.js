import React from 'react'
import {shallow} from 'enzyme'
import {expenses} from '../fixtures/expenses'
import {Create} from '../../components/Create';


test('should handle onsubmit', () => {
    const history = { push:jest.fn() }
    const addExpense = jest.fn()

    const wrapper = shallow(<Create history = {history} addExpense = {addExpense} />)
    expect(wrapper).toMatchSnapshot()
}) 

//checking if both our spies are called with correct stuff

test('should handle onsubmit', () => {
    const history = { push:jest.fn() }
    const addExpense = jest.fn()

    const wrapper = shallow( <Create history = {history} addExpense = {addExpense} /> )
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]) // after this line we can check if all our spyies were called with correct stuff, because we call onSubmit here
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(addExpense).toHaveBeenLastCalledWith(expenses[1])
}) 