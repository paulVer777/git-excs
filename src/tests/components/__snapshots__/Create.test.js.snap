// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle onsubmit 1`] = `
<div>
  <h1>
    Create page
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
